src/k2/tso/service/Service.cpp

```diff
@@ -0,0 +1,220 @@
+    return RPCResponse(Statuses::S200_OK("OK"), std::move(response));
+}
+
+seastar::future<std::tuple<Status, dto::GetTimestampResponse>>
+TSOService::_handleGetTimestamp(dto::GetTimestampRequest&& request) {
+    (void)request; // nothing is passed in
+    if (_tsoId == 0) {
+        return RPCResponse(Statuses::S410_Gone("this server is not authorized to generate timestamps"), dto::GetTimestampResponse{});
+    }
+    _getGPSNow();
+    _timestampErrors.add(nsec(_lastGPSTime.error).count());
+
+    if (_lastGPSTime.error > _errorBound()/2) {
+        // no need to bother doing anything else - the error in gps is too high
+        K2LOG_W(log::tsoserver, "large gps error detected: {}", _lastGPSTime);
+        return RPCResponse(Statuses::S503_Service_Unavailable("gps error too high at the moment"), dto::GetTimestampResponse{});
+    }
+    // We now have to map a GPS timepoint (real +-error) to a K2 Timestamp([endCount-delta: endCount]).
+    // Although gps.real is a monotonically-increasing value, we cannot simply assign it as endCount since
+    // the true time may be bigger than gps.real (e.g. gps.real + error).
+    // Instead, we produce a timestamp which is guaranteed to contain gps.real +- gps.error.
+    // GPS:           |          |           |
+    //                 <error> <real> <error>
+    // TS:      |                                       |
+    //              <delta>                     <endCount>
+    // The requirement for the timestamp then become:
+    // 0. endCount > lastEndCount           // strictly-increasing
+    // 1. real + error <= endCount          // contain upperbound of gps time
+    // 2. real - error >= endCount - delta  // contain lowerbound of gps time
+    // 3. delta <= errorBound               // the error in the timestamp is no bigger than our advertised error bound
+    //
+    // To achieve this, we use the fact that we're allowed to produce all of our timestamps with constant max error
+    uint64_t delta = nsec(_errorBound()).count(); // condition #3: error is no-greater than error bound
+    // and then we use the monotonic gps.real value, shifting it by a constant so that it remains monotonic
+    uint64_t endCount = nsec(_lastGPSTime.real).count() + delta/2;
+    // Thus Timestamp([endCount - error, error]) is guaranteed to include the entirety of the gps time
+
+    // condition #0: now we have to make sure that endCount is strictly-increasing
+    // FYI, probably the GPS timestamp did not update across consecutive calls to TSO.
+    if (endCount <= _lastGeneratedEndCount) {
+        // increment the last returned value and use it as the end count
+        ++_lastGeneratedEndCount;
+        endCount = _lastGeneratedEndCount;
+    }
+
+    // ensure the result is a valid TSO timestamp within error bound
+    if ((uint64_t)nsec(_lastGPSTime.real - _lastGPSTime.error).count() < endCount - delta ||
+        (uint64_t)nsec(_lastGPSTime.real + _lastGPSTime.error).count() > endCount) {

```

Comment:
Shouldn't this condition always be true if your check above (_lastGPSTime.error > _errorBound()/2) passes?